core/web/resolver/mutation.go

```diff
@@ -462,3 +462,123 @@ func (r *Resolver) DeleteVRFKey(ctx context.Context, args struct {
 
 	return NewDeleteVRFKeyPayloadResolver(key, nil), nil
 }
+
+func (r *Resolver) ApproveJobProposal(ctx context.Context, args struct {
+	ID graphql.ID
+}) (*ApproveJobProposalPayloadResolver, error) {
+	jp, err := r.executeJobProposalAction(ctx, jobProposalAction{
+		args.ID, approve, nil,
+	})
+
+	if err != nil {
+		if errors.Is(err, sql.ErrNoRows) {
+			return NewApproveJobProposalPayload(nil, err), nil
+		}
+
+		return nil, err
+	}
+
+	return NewApproveJobProposalPayload(jp, nil), nil
+}
+
+func (r *Resolver) CancelJobProposal(ctx context.Context, args struct {
+	ID graphql.ID
+}) (*CancelJobProposalPayloadResolver, error) {
+	jp, err := r.executeJobProposalAction(ctx, jobProposalAction{
+		args.ID, cancel, nil,
+	})
+
+	if err != nil {
+		if errors.Is(err, sql.ErrNoRows) {
+			return NewCancelJobProposalPayload(nil, err), nil
+		}
+
+		return nil, err
+	}
+
+	return NewCancelJobProposalPayload(jp, nil), nil
+}
+
+func (r *Resolver) RejectJobProposal(ctx context.Context, args struct {
+	ID graphql.ID
+}) (*RejectJobProposalPayloadResolver, error) {
+	jp, err := r.executeJobProposalAction(ctx, jobProposalAction{
+		args.ID, reject, nil,
+	})
+
+	if err != nil {
+		if errors.Is(err, sql.ErrNoRows) {
+			return NewRejectJobProposalPayload(nil, err), nil
+		}
+
+		return nil, err
+	}
+
+	return NewRejectJobProposalPayload(jp, nil), nil
+}
+
+func (r *Resolver) UpdateJobProposalSpec(ctx context.Context, args struct {
+	ID    graphql.ID
+	Input *struct{ Spec string }
+}) (*UpdateJobProposalSpecPayloadResolver, error) {
+	jp, err := r.executeJobProposalAction(ctx, jobProposalAction{
+		args.ID, updateSpec, args.Input,
+	})
+
+	if err != nil {
+		if errors.Is(err, sql.ErrNoRows) {
+			return NewUpdateJobProposalSpecPayload(nil, err), nil
+		}
+
+		return nil, err
+	}
+
+	return NewUpdateJobProposalSpecPayload(jp, nil), nil
+}
+
+type jobProposalAction struct {
+	jpID   graphql.ID
+	name   JobProposalAction
+	update *struct{ Spec string }
+}
+
+func (r *Resolver) executeJobProposalAction(ctx context.Context, action jobProposalAction) (*feeds.JobProposal, error) {
```

Comment:
I'm not a fan of a generic helper function which handles all of this. If we did this, whats to stop us from putting every single mutation as an action, and type switching for different logic. 

Defining the logic in the resolver seems much more readable than this. While I could see a use case to make the Approve/Cancel/Reject use a helper function,  the UpdateSpec is one that has different behaviour than a state change and should be treated differently.

Reply:
Then I'll move the UpdateSpec out of the helper. I added it to the helper since the code structure in the controller was the same for UpdateSpec as it was for the other actions of a Job Proposal.